ksql-engine/src/main/java/io/confluent/ksql/engine/Checksum.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright 2019 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"; you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.engine;
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import io.confluent.ksql.statement.ConfiguredStatement;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * This class represents data that provides enough information to identify
+ * whether two instances of {@link io.confluent.ksql.KsqlExecutionContext}
+ * are equivalent.
+ *
+ * <p>Today, it is represented as a chain of hashes of all statements that
+ * were successfully executed. Any other information that may cause engines
+ * to behave differently should be added to this checksum.</p>
+ */
+@JsonIgnoreProperties(ignoreUnknown = true)
+public final class Checksum {
+
+  private static final Checksum NO_CHECKSUM = new Checksum(-1);
```

Comment:
The way we're using NO_CHECKSUM feels awkward to me. At a high level, I think what we want is to either enable or disable checksums at the statement level. If its enabled (new statements that aren't set or unset), do the verification and incorporate the statement into the current checksum, and if its disabled (old statements or set/unset), don't verify and don't incorporate the statement. IMO this would be better expressed by making Checksum optional in ConfiguredStatement, and making the decision in that class.